core/templates/dev/head/collection_editor/editor_tab/CollectionLinearizerServiceSpec.js

```diff
@@ -0,0 +1,204 @@
+          expect(
+              CollectionLinearizerService.getCollectionNodesInPlayableOrder(
+                  collection)).toEqual([]);
+        });
+
+    it('should correctly remove the first node from a collection', function() {
+      var collection = _createLinearCollection();
+      CollectionLinearizerService.removeCollectionNode(collection, 'exp_id0');
+      expect(collection.containsCollectionNode('exp_id0')).toBe(false);
+      expect(
+          CollectionLinearizerService.getCollectionNodesInPlayableOrder(
+              collection)).toEqual([secondCollectionNode, thirdCollectionNode]);
+    });
+
+    it('should correctly remove the last node from a collection', function() {
+      var collection = _createLinearCollection();
+      CollectionLinearizerService.removeCollectionNode(collection, 'exp_id2');
+      expect(collection.containsCollectionNode('exp_id2')).toBe(false);
+      expect(
+          CollectionLinearizerService.getCollectionNodesInPlayableOrder(
+              collection)).toEqual([firstCollectionNode, secondCollectionNode]);
+    });
+
+    it('should correctly remove a middle node from a collection', function() {
+      var collection = _createLinearCollection();
+      CollectionLinearizerService.removeCollectionNode(collection, 'exp_id1');
+      expect(collection.containsCollectionNode('exp_id1')).toBe(false);
+      expect(
+          CollectionLinearizerService.getCollectionNodesInPlayableOrder(
+              collection)).toEqual([firstCollectionNode, thirdCollectionNode]);
+    });
+  });
+
+  describe('Collection linearizer node append function', function() {
+    it('should correctly append a node in an empty collection', function() {
+      var collection = CollectionObjectFactory.createEmptyCollection();
+      CollectionLinearizerService.appendCollectionNode(collection, 'exp_id0',
+          firstCollectionNode.getExplorationSummaryObject());
+      expect(
+          CollectionLinearizerService.getCollectionNodesInPlayableOrder(
+                  collection)).toEqual([firstCollectionNode]);
+    });
+
+    it('should correctly append a node in a non-empty collection', function() {
+      var collection = CollectionObjectFactory.createEmptyCollection();
+      collection.addCollectionNode(firstCollectionNode);
+      collection.addCollectionNode(secondCollectionNode);
+      expect(
+          CollectionLinearizerService.getCollectionNodesInPlayableOrder(
+              collection)).toEqual([firstCollectionNode, secondCollectionNode]);
+      CollectionLinearizerService.appendCollectionNode(collection, 'exp_id2',
+          thirdCollectionNode.getExplorationSummaryObject());
+      expect(
+          CollectionLinearizerService.getCollectionNodesInPlayableOrder(
+              collection)).toEqual([firstCollectionNode, secondCollectionNode,
+                  thirdCollectionNode]);
+    });
+  });
+
+  describe('Collection linearizer node shifting function', function() {
```

Comment:
You're right that this is testing some functionality that isn't exposed to the user. However, the ``CollectionLinearizerService`` doesn't promise it will be used in the UI layer. It's just a general purpose utility that happens to be used by a directive. It certainly could, in the future, used by some other general-purpose service which automatically generates collections. We want to make sure the policy of the ``CollectionLinearizerService`` is well-defined and well-tested in case future developers try using it in ways the current UI does not.